Add functionality to edit channel topic from the user interface

ollipa opened the pull request:
Closes #277.

This PR adds the functionality to edit channel topic from the user interface. The user can either double-click the topic to toggle editing or select "Edit topic" from the channel context menu. The new topic is saved when the user presses enter. Editing is canceled and the changes are discarded if the input field loses focus or if the user presses esc key.

![image](https://user-images.githubusercontent.com/25169984/62426746-810d0b00-b6f1-11e9-95a1-57e012ef31f0.png)

![image](https://user-images.githubusercontent.com/25169984/62426748-866a5580-b6f1-11e9-96da-62b5d5af5434.png)

I've been following this project for quite a while and always thought that I will make a contribution some day. I hope this is helpful :)

xPaw commented on client/components/Chat.vue:
```diff
@@ -118,6 +129,24 @@ export default {
+			this.channel.editTopic = true;
+
+			this.$nextTick(() => {
+				$(`#chan-${this.channel.id} .topic-edit`).focus();
```
We want to eventually remove jquery, would be nice to do with vanilla js

xPaw commented:
Thanks for the PR, appears to work fine as you'd expect it to. I think there should be a save button (checkmark on the right for example).

xPaw commented on client/components/Chat.vue:
```diff
@@ -118,6 +129,24 @@ export default {
+			this.channel.editTopic = false;
+			const newTopic = event.target.value;
+
+			if (newTopic && this.channel.topic !== newTopic) {
```
This check prevents you from removing a topic.

ollipa replied:
I actually added that check after I noticed that sending only `/topic` just prints out the current topic in the channel. I could not find a standard way of clearing topic in IRC. However, I could set the topic equal to one white-space character if the user sends an empty topic. What do you think?

xPaw replied:
That's a good point. Leave it as is then.

xPaw replied:
Edit, apparently it's possible by doing `> TOPIC #ircv3 :`, but that also requires changing the server command to work with this.

xPaw replied:
I created #3351 to handle this

xPaw replied:
Can you change this to send `/raw TOPIC ${channelname} :${newTopic}` command?

I just don't know if `/topic ` is a good idea to clear the topic. #3352 fixes the raw command so the colon works as expected.

ollipa replied:
I changed this to use raw command and tested that it works as expected with your change.

xPaw commented on client/components/Chat.vue:
```diff
@@ -118,6 +129,24 @@ export default {
 		hideUserVisibleError() {
 			this.$root.currentUserVisibleError = null;
 		},
+		editTopic() {
```
This should have a check for channel type, right now you can double click in lobby (and other types) and the input field appears.

xPaw commented on client/css/style.css:
```diff
@@ -978,6 +978,17 @@ background on hover (unless active) */
 }
 
+#windows .header .topic-edit {
+	color: #222;
```
Doing the following styles makes it work nicely on other themes (e.g. with dark background):

```css
    color: inherit;
    background: inherit;
```

ollipa commented:
> Thanks for the PR, appears to work fine as you'd expect it to. I think there should be a save button (checkmark on the right for example).
> 
> EDIT: Any ideas how this will work on mobile due to very limited width?

It actually works quite well at least on Android & Chrome. The input field automatically scrolls if the topic gets too long. See the screenshots below.

![image](https://user-images.githubusercontent.com/25169984/62479048-440a4c80-b7b5-11e9-9d6c-42d5a617ce1b.png)

![image](https://user-images.githubusercontent.com/25169984/62479113-5f755780-b7b5-11e9-877b-2b8f86b22b7c.png)

Obviously it is not the best user experience because of the limited space. I was thinking I could later try work on issue #276 and come up with a solution that would improve the topic editing as well on mobile.

I can add the save button, but I think it would be better to hide it in the mobile view because of limited space.

ollipa commented:
@xPaw, I have now addressed all your comments. I'll check how to add the save button next.

xPaw commented:
That `/raw` PR was merged, you can rebase your PR so it can be tested correctly.

ollipa commented:
I rebased and added a save button to the edit topic input field:
![image](https://user-images.githubusercontent.com/25169984/62834728-1b70cf80-bc59-11e9-9edf-8d0389b0d4fa.png)

Previously the editing was canceled if the edit topic input field lost focus (clicking anywhere outside of the input box). I could not figure out a simple way to make the save button work together with that functionality so I removed it. Editing can still be canceled by hitting esc key or just saving the topic without making any changes.

xPaw commented:
Thanks for the PR! Anything you want to do before merge?

> I could not figure out a simple way to make the save button work together with that functionality so I removed it.

What trouble was it causing?

xPaw commented on client/css/style.css:
```diff
@@ -978,6 +984,42 @@ background on hover (unless active) */
+}
+
+.topic-container .save-topic span {
+	font: normal normal normal 14px/1 FontAwesome;
```
Look on line 249, we define the fontawesome font in the big selector there.

ollipa replied:
This is fixed now.

ollipa commented:
> Thanks for the PR! Anything you want to do before merge?
> 

I think this is ready to merge. I've tested the changes myself with Chrome, Firefox & Chrome on Android and everything seems to work as intended.

> > I could not figure out a simple way to make the save button work together with that functionality so I removed it.
> 
> What trouble was it causing?

Earlier I had a Vue event handler on the input field, which canceled the editing when the input field lost focus:
```
@blur="channel.editTopic = false"
```
After adding the save button, this event handler was triggered also when the save button was clicked, causing the changes to be discarded instead of being saved.

xPaw commented:
Perhaps nextTick on blur could help?

ollipa commented:
> Perhaps nextTick on blur could help?

I tried that but it didn't work.

xPaw commented:
Oh well, problem for another day.
